### PR TITLE
Don't call mbedtls_cipher_setkey twice

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
The documentation doesn't explicitly say whether it's allowed or not.
This currently works with the default software implementation, but
only by accident. It isn't guaranteed to work with new ciphers or with
alternative implementations of individual ciphers, and it doesn't work
with the PSA wrappers. So don't do it.

Alternative to https://github.com/ARMmbed/mbedtls/pull/2581

Follow-up that documents that this is forbidden and add a check and a test: https://github.com/ARMmbed/mbedtls/pull/2584

### CODE REVIEW

File: tests/suites/test_suite_cipher.function [hunk 1]
```diff
@@ -1011,6 +1011,20 @@ void auth_crypt_tv( int cipher_id, data_t * key, data_t * iv,
     TEST_ASSERT( memcmp( output, clear->x, clear->len ) == 0 );
 
     /* then encrypt the clear->x and make sure we get the same ciphertext and tag->x */
+    TEST_ASSERT( mbedtls_cipher_reset( &ctx ) == 0 );
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( use_psa == 1 )
+    {
+        TEST_ASSERT( 0 == mbedtls_cipher_setup_psa( &ctx,
```
Comment (yanesca) [inline tests/suites/test_suite_cipher.function]: Don't we need to do an `mbedtls_cipher_free()` before this?

**Comment by Patater** [comment]
When I went to look at how reset works, I found something interesting. It looks like we don't support reseting PSA-backed ciphers yet, either.

```
int mbedtls_cipher_reset( mbedtls_cipher_context_t *ctx )
{
    CIPHER_VALIDATE_RET( ctx != NULL );
    if( ctx->cipher_info == NULL )
        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );

#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( ctx->psa_enabled == 1 )
    {
        /* We don't support resetting PSA-based
         * cipher contexts, yet. */
        return( MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE );
    }
#endif /* MBEDTLS_USE_PSA_CRYPTO */

    ctx->unprocessed_len = 0;

    return( 0 );
}
```

File: tests/suites/test_suite_cipher.function (hunk 1 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_cipher.function]: Oh. So `mbedtls_cipher_reset` doesn't do what I expected at all. It allows the context to be reused with the same key, same direction, and _same IV_. When is this ever useful?

I'll replace the call to reset() by a call to free().

Comment (yanesca) [reply tests/suites/test_suite_cipher.function]: Indeed, it sounds very dangerous, and the documentation is not very helpful. At least we always call `set_iv` first in the library.

**Pull request merged**